### DISCUSSION TITLE
Hermes as default engine

### DISCUSSION THREAD

**Discussion by cortinico** [reactwg/react-native-new-architecture#4] [Deep Dive]
Hi all,
I'm opening this discussion to discuss the implication of having [Hermes](https://github.com/facebook/hermes) as **default engine for the New RN Architecture**.

As part of the New Architecture Rollout, we're going to set Hermes as the default engine for all the new React Native applications. It will still be possible to use other JS engines with your application, but we expect to develop new features and provide better support having a Hermes-first approach.

Feel free to ask your questions/ideas/doubts related to this in this discussion.

## Relevant resources

- [Toward Hermes being the Default](https://reactnative.dev/blog/2021/10/26/toward-hermes-being-the-default)

**Comment by cortinico** [comment]
# Bundled Hermes on Android

Hi all,
I'd like to share a change that we recently shipped, that is going to impact the way how Hermes is consumed on Android. This is part of the effort of [making Hermes the default engine for React Native](https://reactnative.dev/blog/2022/03/15/an-update-on-the-new-architecture-rollout#towards-hermes-as-default-engine).

## What has changed

We changed the way how Hermes is built and distributed. Starting with React Native 0.69.0, every version of React Native will be built alongside to a Hermes version.

This will make sure that the version of Hermes that has been built when RN was released, is **fully compatible** with that RN version.

Historically, we had problems with matching versions of Hermes with versions of RN. This fully eliminates this problem and will ideally offer a bundled JS engine that is compatible with the specific RN version.

## How this will impact me

On Android, we're going to update the [`android/app/build.gradle`](https://github.com/facebook/react-native/blob/main/template/android/app/build.gradle) file in the default template the following way:

```diff
dependencies {
    // ...

    if (enableHermes) {
+       implementation("com.facebook.react:hermes-engine:+") {
+           exclude group:'com.facebook.fbjni'
+       }
-       def hermesPath = "../../node_modules/hermes-engine/android/";
-       debugImplementation files(hermesPath + "hermes-debug.aar")
-       releaseImplementation files(hermesPath + "hermes-release.aar")
    } else {
        implementation jscFlavor
    }
}
```

Practically speaking, you won't be consuming `hermes-debug.aar` and `hermes-release.aar` from the `hermes-engine` NPM package anymore. We're probably going to [remove the dependency](https://github.com/facebook/react-native/blob/c418bf4c8fe8bf97273e3a64211eaa38d836e0a0/package.json#L105) on `hermes-engine` entirely in one of the future version.

## Why this has changed

As mentioned, we're not building a version of Hermes together with React Native.

Both Hermes and React Native, share the JSI code ([here](https://github.com/facebook/hermes/tree/main/API/jsi/jsi) and [here](https://github.com/facebook/react-native/tree/main/ReactCommon/jsi/jsi)). 
If the two JSI copies of JSI gets out of sync, a build of Hermes won't be compatible with a build of React Native. You can read more about this problem, ABI incompatibility, here:
- https://github.com/react-native-community/discussions-and-proposals/issues/257

To overcome this problem, we're now downloading the Hermes source code at build time, and making sure only one copy of JSI is used when building Hermes. 

## How can I test this right now if I wish?

The latest nightly of React Native ([0.0.0-20220316-2009-8200f9159](https://www.npmjs.com/package/react-native/v/0.0.0-20220316-2009-8200f9159)), contains a bundled version of Hermes.

You shoudl be able to test it with:
1. `npx react-native init RNHERMES --version 0.0.0-20220316-2009-8200f9159`
1. Set `enableHermes` to true in your `android/app/build.gradle`
1. Edit the `dependencies` block in `android/app/build.gradle` as mentioned above.

Feel free to report any bug you might face, your feedback is deeply appreciated.

## Can I still use another engine?

Yes, you can still use the `enableHermes` variable to control which JS engine is picked up. This change will impact only **how Hermes is built and bundled** for you.

Please note that, at the time of writing, the default for `enableHermes` is `false`. We're looking into updating the template to have it default to `true` in the near future.

## Information for Contributors 

If you're a React Native contributor or you're building an extension for Hermes, please read further as we expalin how this mechanism works.

If you're a user of React Native, those information won't be relevant for you.

### How is this mechanism working?

This mechanism relies on downloading a tarball from the facebook/hermes repository inside the facebook/react-native repository. We have a similar mechanism in place for other dependencies (Folly, Glog, etc.)

This allows us to "similate" a monorepo setup in the OpenSource space.

### How is the version of Hermes picked?

When building React Native from `main`, we will be fetching a tarball of `main` of facebook/hermes and building it as part of the build process of React Native.

When cuttin a release branch for RN, we will be tagging a version of Hermes(with this GH Action https://github.com/facebook/hermes/blob/main/.github/workflows/create-tag.yml).

This tag name will then be stored inside the `sdks/.hermesversion` file inside RN in the release branch. We will then be using this tag name to download the specific tarball. This is essentially the synchronization point between the two repository. 

Moreover, this approach allows us to tag a subsequent version of Hermes, should we need a patch once we're inside the RN release process.

In a sense, you can think of this approach similarly to a git submodule.

If you're building on top of Hermes, you can rely on those tags to understand which version of Hermes was used when building React Native.

### I'm afraid this change is impacting me.

I'd like to stress that this is essentially an organizational change on "where" Hermes is built and "how" the code is syncronized between the two repos. The change should be fully transparent to our users.

Historically, we used to cut a release of Hermes for a specific version of RN.
For instance: https://github.com/facebook/hermes/releases/tag/v0.11.0
was version `v0.11.0 for RN0.68.x`. 

In the new approach, you will instead have a tag that will represent the version of RN that was cut.

Feel free to share your feedback and opinion in the thread below

### Android implementation details

To implement this on Android, we've added a new build inside the `/ReactAndroid/hermes-engine` of React Native that will take care of building Hermes and packaging for consumption.

See here for more context:
- https://github.com/facebook/react-native/pull/33396

You can now fire a build of Hermes engine by invoking:

```
./gradlew :ReactAndroid:hermes-engine:assembleDebug
```

(this will create a debug build of Hermes).

Please note that now you'll need to have installed `cmake`, `ninja` and `python3` in your machine in order to build React Android (i.e. if you'are a contributor to React Native).

If you're wondering why the:

> exclude group:'com.facebook.fbjni'

Line is needed. React Native is consuming fbjni using the non-prefab approach (i.e. unzipping the .aar). Hermes-engine, and other libraries, are using prefab instead to consume fbjni. We're looking into [fixing this here](https://github.com/facebook/react-native/pull/33397) so the Hermes import will be a oneliner.

If you're wondering why we moved from `releaseImplementation`, `debugImplementation` to just `implementation`: it's because the newer hermes-engine is variant aware and will proper match a debug build of the engine with a debug build of your app, and so on. You don't need any custom configuration here (even if you use `staging` or other build flavors).

-----

If you have any questions/doubts on the approach, feel free to ask here 👍 

EDIT: Added the "Information for Contributors" section

**Reply by douglowder** [reply]
This is great to hear! I have a couple of questions on this:

- Any changes in how rn-tester is built for Android? I'm working on the TV fork and seeing some compile issues there in 0.68.
- Will Hermes also be default for iOS? I'm going to have to make sure that the non-Hermes build works fully because there is no Hermes for tvOS (arm64 arch) at the moment.

**Reply by cortinico** [reply]
> * Any changes in how rn-tester is built for Android? I'm working on the TV fork and seeing some compile issues there in 0.68.

Yes, namely we're going to update the `packages/rn-tester/android/app/build.gradle` file as follows:

```diff
dependencies {
    // ...
+   hermesImplementation(":ReactAndroid:hermes-engine")
-   def hermesPath = '$projectDir/../../../../../node_modules/hermes-engine/android/'
-   hermesDebugImplementation files(hermesPath + "hermes-debug.aar")
-   hermesReleaseImplementation files(hermesPath + "hermes-release.aar")
```

You could try it right now if you clone `react-native` and build RN Tester locally.

> * Will Hermes also be default for iOS? I'm going to have to make sure that the non-Hermes build works fully because there is no Hermes for tvOS (arm64 arch) at the moment.

Yes, we're going to share more in the coming future about Hermes & iOS.

**Reply by kraenhansen** [reply]
Thanks for sharing this. I think this makes a lot of sense. I imagine this will also make it easier to test changes to Hermes in the context of a React Native app.
Library maintainers shouldn't really have to worry about this if they link against JSI and don't have to link with Hermes directly.

A bit related: We're are faced with a similar issue which [we need to address at some point](https://github.com/realm/realm-js/pull/4027): We're currently pre-building our binary, but can't actually rely on the JSI ABI to remain stable across React Native versions.

**Reply by NickGerleman** [reply]
The set of tarballs is a projection of the state of the Hermes OSS repo. **Q: is projection always along Hermes main branch?** 

There is a separate model of repository projection used by Yoga. It exists as its own GitHub repo, but also is projected into the C++ source tree of the `react-native` repo (and then npm package). See https://github.com/facebook/react-native/tree/main/ReactCommon/yoga

I am unfamiliar about how Meta internally has this projection set up, but projecting source to the repo has some properties that I think would work well here. It means that React Native's existing repo-versioning would allow servicing cherry-pick style changes to Hermes sources, at the same cadence as RN itself. It would also allow RN users/redistributors (like RNW) to leverage their existing systems to build cryptographically signed binaries from RN sources.

Downloading unique tarballs or zip files during the build is a reality we have, but I've seen them often have a negative impact to build determinism or performance compared to rolling the sources into an already used package manager. E.g. inlining into the C++ source distributed as part of the existing NPM package, or separating into published artifacts. I'm curious why the format was chosen compared to alternatives.

**Reply by cortinico** [reply]
> Q: is projection always along Hermes main branch?

Sorry I forgot to answer this. The answer is no. I've update my comment with a section for contributors that explains the setup in details and clarifies (hopefully) most of the doubts. Let me know if it's not clear.

**Comment by cortinico** [comment]
Hey all,
I'd like to share that we published an updated deep-dive on how Bundled Hermes works for both Android & iOS on the React Native website here: https://reactnative.dev/architecture/bundled-hermes
